misc/python/materialize/cloudtest/k8s/environmentd.py

```diff
@@ -0,0 +1,168 @@
+            V1EnvVar(name="AWS_ACCESS_KEY_ID", value="minio"),
+            V1EnvVar(name="AWS_SECRET_ACCESS_KEY", value="minio123"),
+        ]
+
+        ports = [V1ContainerPort(container_port=5432, name="sql")]
+
+        volume_mounts = [
+            V1VolumeMount(name="data", mount_path="/data"),
+            V1VolumeMount(name="secrets-mount", mount_path="/secrets"),
+        ]
+
+        s3_endpoint = urllib.parse.quote("http://minio-service.default:9000")
+
+        container = V1Container(
+            name="environmentd",
+            image=self.image("environmentd"),
+            args=[
+                "--storaged-image=" + self.image("storaged"),
+                "--computed-image=" + self.image("computed"),
+                f"--persist-blob-url=s3://minio:minio123@test/test?endpoint={s3_endpoint}&region=minio",
+                "--orchestrator=kubernetes",
+                "--orchestrator-kubernetes-image-pull-policy=never",
+                "--persist-consensus-url=postgres://postgres@postgres.default?options=--search_path=consensus",
+                "--adapter-stash-url=postgres://postgres@postgres.default?options=--search_path=catalog",
+                "--storage-stash-url=postgres://postgres@postgres.default?options=--search_path=storage",
+                "--unsafe-mode",
+            ],
+            env=env,
+            ports=ports,
+            volume_mounts=volume_mounts,
+        )
+
+        volumes = [
+            V1Volume(
+                name="secrets-mount",
+                secret=V1SecretVolumeSource(
+                    default_mode=400, secret_name="user-managed-secrets"
+                ),
+            )
+        ]
+        pod_spec = V1PodSpec(containers=[container], volumes=volumes)
+        template_spec = V1PodTemplateSpec(metadata=metadata, spec=pod_spec)
+        claim_templates = [
+            V1PersistentVolumeClaim(
+                metadata=V1ObjectMeta(name="data"),
+                spec=V1PersistentVolumeClaimSpec(
+                    access_modes=["ReadWriteOnce"],
+                    resources=V1ResourceRequirements(requests={"storage": "1Gi"}),
+                ),
+            )
+        ]
+
+        self.stateful_set = V1StatefulSet(
+            api_version="apps/v1",
+            kind="StatefulSet",
+            metadata=metadata,
+            spec=V1StatefulSetSpec(
+                service_name="environmentd",
+                replicas=1,
+                pod_management_policy="Parallel",
+                selector=label_selector,
+                template=template_spec,
+                volume_claim_templates=claim_templates,
+            ),
+        )
+
+    def host_port(self) -> Tuple[str, str]:
+        url = subprocess.check_output(
+            ["minikube", "service", "environmentd", "--url"]
+        ).decode("ascii")
+        parts = str(url).split(":")
+        host = parts[1][2:]
+        port = parts[2].strip()
+        return (host, port)
```

Comment:
This seems a little fishy, because the method on the `StatefulSet` is referencing the `Service`! I recommend introducing a new class called `K8sApplication(K8sResource)` that simply groups together a number of other resources. Then you could have an `EnvironmentdApplication` that depends on both the statefulset and the service, and it'd be a more natural home for these methods, IMO.

Reply:
Thanks, good catch. I have moved all of that code over to the `EnvironmentdService` . I have opened a follow-up ticket for introducing an `Application` abstraction. https://github.com/MaterializeInc/materialize/issues/13720